autowarefoundation/autoware.universe
# feat: use `pose_source` and `twist_source` for selecting localization methods

THIS PR SHOULD BE MERGED AFTER A CERTAIN DISCUSSION IN https://github.com/orgs/autowarefoundation/discussions/3659 

## Description
Change localization_launch architecture so that we can select localization methods with 
- `pose_source`: `ndt`, `yabloc`, or `eagleye`
- `twist_source`: `gyro_odom`, or `eagleye`

You can specify these now like as follow:
```bash
$ ros2 launch autoware_launch autoware.launch.xml ... pose_source:=yabloc twist_source:=gyro_odom ...
```

NOTE: This PR does NOT (and must not) change the performance of any localization mode. 


## Related links
- https://github.com/autowarefoundation/autoware_launch/pull/442
- https://github.com/autowarefoundation/autoware-documentation/pull/418


INTERNAL LINK: https://star4.slack.com/archives/CEL0DR2S2/p1688988460478459
## Tests performed

- For logging simulator:
  - Run with default launch command to verify that it works
  - Run with `pose_source:=yabloc` with yabloc sample rosbag
  - Run with `pose_source:=ndt_scan_matcher twist_source:=eagleye` to confirm that eagleye with twist_estimator mode works
  - Run with `pose_source:=eagleye twist_source:=eagleye` to confirm that eagleye with pose_twist_estimator mode works
- For planning simulator:
  - Run simple planning simulator and confirmed that it engages and starts driving
  - Run scenario based planning simulation based on [a tutorial](https://autowarefoundation.github.io/autoware-documentation/main/tutorials/scenario-simulation/planning-simulation/scenario-test-simulation/) and confirmed that the vehicle engages and starts driving

## Notes for reviewers

Some large modifications have been made in this PR, including:

- localization_launch directory structure
  - `pose_estimator` and `twist_estimator` has been merged into `pose_twist_estimator`
  - `map4_localization_launch` has been removed
- YabLoc
  - Namespace has been modified (`/localization/yabloc/...` to `/localization/pose_estimator/...`)

For testing this PR, please use the following dataset:
- NDT: any data with LiDAR is OK (e.g. sample-map-rosbag)
- YabLoc: please find the YabLoc sample dataset [in this link](https://github.com/autowarefoundation/autoware.universe/pull/3946)
- Eagleye: Sample data is NOT sufficient, so please use some arbitrary longer data, e.g. from JPN TAXI (eagleye takes a lot of time for initialization)



## Interface changes
- Some of the YabLoc internal topic names are modified, but should not have any effect to other packages (No influence when using other pose estimator such as `ndt_scan_matcher`)



## Effects on system behavior
None


## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.


## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.


## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.


After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/
